syz-manager/covfilter.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"regexp"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/mgrconfig"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+type CoverFilter struct {
+	pcStart     uint32
+	pcSize      uint32
+	pcEnd       uint32
+	weightedPCs map[uint32]float32
+
+	bitmapFilename string
+	target         *targets.Target
+}
+
+func createCoverageFilter(cfg *mgrconfig.Config, target *targets.Target) (covFilterFilename string, err error) {
+	files := cfg.CovFilter.Files
+	funcs := cfg.CovFilter.Functions
+	rawPCs := cfg.CovFilter.RawPCs
+	if len(files) == 0 && len(funcs) == 0 && len(rawPCs) == 0 {
+		return "", nil
+	}
+	filesRegexp, err := getRegexps(files)
+	if err != nil {
+		return "", err
+	}
+	funcsRegexp, err := getRegexps(funcs)
+	if err != nil {
+		return "", err
+	}
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		log.Logf(0, "initialize coverage information...")
+		if err = initCover(target, cfg.KernelObj, cfg.KernelSrc, cfg.KernelBuildSrc); err != nil {
+			return "", err
+		}
+	}
+
+	covFilter := CoverFilter{
+		weightedPCs:    make(map[uint32]float32),
+		target:         target,
+		bitmapFilename: cfg.Workdir + "/" + "syz-cover-bitmap",
+	}
+	err = covFilter.initWeightedPCs(filesRegexp, funcsRegexp, rawPCs)
+	if err != nil {
+		return "", err
+	}
+	err = covFilter.createBitmap()
+	if err != nil {
+		return "", err
+	}
+	return covFilter.bitmapFilename, nil
+}
+
+func (covFilter *CoverFilter) initWeightedPCs(filesRegexp, funcsRegexp []regexp.Regexp, rawPCsFiles []string) error {
+	for _, f := range rawPCsFiles {
+		rawFile, err := os.Open(f)
+		if err != nil {
+			return fmt.Errorf("failed to open raw PCs file: %v", err)
+		}
+		for {
+			var encode uint64
+			_, err := fmt.Fscanf(rawFile, "0x%x\n", &encode)
+			if err != nil {
+				break
+			}
+			pc := uint32(encode & 0xffffffff)
+			weight := float32((encode >> 32) & 0xffff)
+			covFilter.weightedPCs[pc] = weight
+		}
+		rawFile.Close()
+	}
+
+	enabledFuncs := make(map[string]bool)
+	enabledFiles := make(map[string]bool)
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		if reportGenerator == nil {
+			return fmt.Errorf("reportGenerator used without initialization")
+		}
+		pcs := make([]uint64, 0)
+		symNames, cuNames, symPCs := reportGenerator.GetSymbolsInfo()
+		for i, symName := range symNames {
+			for _, r := range funcsRegexp {
+				if ok := r.MatchString(symName); ok {
+					enabledFuncs[symName] = true
+					pcs = append(pcs, symPCs[i]...)
+				}
+			}
+			for _, r := range filesRegexp {
+				if ok := r.MatchString(cuNames[i]); ok {
+					enabledFuncs[symName] = true
+					enabledFiles[cuNames[i]] = true
```

Comment:
We maintain enabledFuncs/enabledFiles just for logging. I would just log here without these additional hops.
It may change order of logging, but it will only affect users who use both funcs and files and I would say it's not too bad and does not worth additional code.

Reply:
Because we traverse all symbols, and a file always includes more than one symbol, that means if we just log here but not trace it by map, enabledFiles would be print out repeatedly.
Or, we can only maintain enabledFiles map and print out enabledFuncs immediately?

Reply:
Right, printing files multiple times is not good.

Thinking more about of this, I think it's also important to detect file/func regexps that don't match anything at all. Highly likely it means a typo. Users will either continue running assuming their filters work when they don't; or users will be puzzled why they are not getting coverage in the target functions. Pretty sure there will be users posting questions on the mailing list asking me, but then there is no easy way to understand what goes wrong based on the logs.

I thinking about something along the following lines:

```
func (covFilter *CoverFilter) initFilesFuncs(filesRegexp, funcsRegexp []*regexp.Regexp, symbols []cover.Symbol) error {
	var enabledFuncs []string
	var enabledFiles []string
	fileDedup := make(map[string]bool)
	used := make(map[*regexp.Regexp]bool)
	for _, sym := range symbols {
		matched := false
		for _, re := range funcsRegexp {
			if re.MatchString(sym.Name) {
				matched = true
				used[re] = true
				enabledFuncs = append(enabledFuncs, sym.Name)
				break
			}
		}
		for _, re := range filesRegexp {
			if re.MatchString(sym.File) {
				matched = true
				used[re] = true
				if !fileDedup[sym.File] = true {
					fileDedup[sym.File] = true
					enabledFiles = append(enabledFiles, sym.File)
				}
				break
			}
		}
		if matched {
			for _, pc := range sym.PCs {
				covFilter.weightedPCs[pc] = 1
			}
		}
	}
	sort.Strings(enabledFuncs)
	sort.Strings(enabledFiles)
	log.Logf(1, "coverage-enabled funcs: %v", enabledFuncs)
	log.Logf(1, "coverage-enabled files: %v", enabledFiles)
	for _, re := range filesRegexp {
		if !used[re] {
			log.Logf(1, "coverage file filter doesn't match anything: %v", re.String())
		}
	}
	for _, re := range funcsRegexp {
		if !used[re] {
			log.Logf(1, "coverage func filter doesn't match anything: %v", re.String())
		}
	}
	// TODO: do we want to error on this? or logging it enough?
	if len(filesRegexp) + len(funcsRegexp) != used {
		return fmt.Errorf("some coverage filters don't match anything")
	}
	return nil
}
```

Reply:
What about printing out which functions or files matched to which regexp?
Users may enable kernel functions base on reading source code, and some kernel functions may become inline functions after compiling. Should this be broken also?
If we print which regexps matched what thing, let users check their regexp by themself.
Also, if users find what matched by a regexp isn't what they want, or just is not totally matched their expectation, they can modify their regexp base on this log.
Do you think it's worth enough to implement things like this?

Reply:
Since my version is already complex enough, let's say we can bear a bit more code to make this more transparent to users.
Perhaps we could use `used := make(map[*regexp.Regexp][]string)` to keep track what regexps matched what and then this will allow us to both print matched strings per-regexp and detect regexps that did not match anything at all.